[CORE 8915] kafka: Implement Consumer Group Lag Metrics

BenPope opened the pull request:
Implement Consumer Group Lag Metrics

For every committed offset, make request to the HWM for the leader of the partition, and then calculate the lag, and expose it in metrics.

## Notes ##
`consumer_group_lag_collection_interval` should be set to the scrape interval.

Currently each shard makes requests, a future optimisation might be to gather the set of topic-partitions for each shard to reduce the number of RPCs.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Features

* metrics: Consumer Group Lag can now be enabled

BenPope committed:
kafka/lag_metrics: Add RPC service

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
kafka/lag_metrics: Add frontend

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
kafka/lag_metrics: Implement RPC service

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
kafka/group_manager: Implement metrics collection

Signed-off-by: Ben Pope <ben@redpanda.com>

IoannisRP commented on src/v/kafka/server/consumer_group_lag_metrics_rpc_types.h:
```diff
@@ -0,0 +1,52 @@
+      partition_offsets_request,
+      serde::version<0>,
+      serde::compat_version<0>> {
+    using rpc_adl_exempt = std::true_type;
```
what does this do?

BenPope replied:
> what does this do?

[Disables support](https://github.com/redpanda-data/redpanda/commit/9814025632b1e4b87aa36537ad646ed72202aebc) for the old serialization code in namespace `reflection`.

IoannisRP commented on src/v/kafka/server/consumer_group_lag_metrics_frontend.cc:
```diff
@@ -0,0 +1,193 @@
+consumer_group_lag_metrics_frontend::get_partition_offsets(
+  partition_offsets_reply req) {
+    partition_offsets_request offsets_req;
+    vlog(
+      klog.warn,
+      "consumer_group_lag_metrics_frontend::get_lag: req: {}",
+      req.data.size());
```
why are all the logs warnings?

are these just for development?

BenPope replied:
> why are all the logs warnings?
> 
> are these just for development?

Yes

IoannisRP commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -127,29 +132,39 @@ ss::future<> group_manager::start() {
+            }
+        });
+    });
+    _lag_metrics_timer.arm(std::chrono::seconds{5});
```
I assume this set to 5s to be easily testable locally, right?

IoannisRP commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1881,4 +1897,79 @@ group_manager::get_group_producers_locally(
+    partition_offsets_reply request;
+
+    using lag = consumer_group_lag_metrics_frontend::lag;
+    struct partition_lag {
+        model::partition_id partition;
+        lag lag;
+    };
```
is this a re-definition of `consumer_group_lag_metrics_frontend::partition_lag`?

BenPope replied:
> is this a re-definition of `consumer_group_lag_metrics_frontend::partition_lag`?

Yeah, a ghost of refactoring past.

BenPope replied:
Done

IoannisRP commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1881,4 +1897,79 @@ group_manager::get_group_producers_locally(
 
+ss::future<> group_manager::collect_consumer_lag_metrics() {
+    vlog(klog.warn, "group_manager::collect_consumer_lag_metrics");
+    partition_offsets_reply request;
```
we create a reply object, fill it up with values that are going to be discarded (the offsets), give it to `get_partition_offsets` which culls it back to `partition_offsets_request` (discarding the offsets) and uses that.

Why don't we make the `partition_offsets_request` directly and pass that to `get_partition_offsets`?

BenPope replied:
> Why don't we make the `partition_offsets_request` directly and pass that to `get_partition_offsets`?

This is refactoring mess.

BenPope replied:
Done

BenPope committed:
kafka/group: Allow group lag metrics to be set

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/f31b9c3442e21d1c64e9f6063ce628e6ea1e53c3..b3fb9ca14429b19eebe60678ada8c6f11cb4b5e6)):
* Remove dead code
* Remove test logging
* Improve log levels
* Use `partition_offsets_reply` where appropriate
* Set the lag metrics
* Change the loop to run every 60s (this is likely to become configurable later)

pgellert commented on src/v/kafka/server/BUILD:
```diff
@@ -155,10 +163,12 @@ redpanda_cc_library(
     ],
     implementation_deps = [
         "//src/v/features:enterprise_feature_messages",
+        "//src/v/serde:named_type",
```
Q1: Should this be a `dep` of `consumer_group_lag_metrics_rpc` instead?
Q2: Is it actually used or is it an unused import?

BenPope replied:
> Q1: Should this be a `dep` of `consumer_group_lag_metrics_rpc` instead?

That didn't seem to help it compile.

> Q2: Is it actually used or is it an unused import?

Compilation fails without it.

pgellert replied:
This compiles for me: https://gist.github.com/pgellert/edb977fab13c5ac4d708adf740d2d3e0. Am I missing something?

BenPope replied:
Done, thanks.

pgellert commented on src/v/kafka/server/consumer_group_lag_metrics_rpc_types.h:
```diff
@@ -0,0 +1,52 @@
+
+    using offsets = chunked_hash_map<
+      model::topic,
+      chunked_hash_map<model::partition_id, kafka::offset>>;
```
nit/question: I'm wondering if the `kafka::offset` should be something more evolvable, eg. a `partition_data` struct. But then it's more efficient to keep it just a `kafka::offset` if we don't end up needing more data there, so I'm fine with the current approach

BenPope replied:
> nit/question: I'm wondering if the `kafka::offset` should be something more evolvable, eg. a `partition_data` struct. But then it's more efficient to keep it just a `kafka::offset` if we don't end up needing more data there, so I'm fine with the current approach

I don't think we need anything else in there.

pgellert commented on src/v/kafka/server/consumer_group_lag_metrics_util.h:
```diff
@@ -0,0 +1,43 @@
+
+namespace kafka {
+
+struct ktp_view {
```
nit: should this go to `model`, next to `ktp.h`?

BenPope replied:
> nit: should this go to `model`, next to `ktp.h`?

I think it could do with a few more features before it's generally usable.

I've moved it to frontend, the only place it's used.

pgellert commented on src/v/kafka/server/consumer_group_lag_metrics_frontend.cc:
```diff
@@ -0,0 +1,135 @@
+    co_return co_await container().map_reduce0(
+      [&ktps](auto& me) {
+          partition_offsets_reply reply;
+          std::ranges::for_each(ktps, [&](const ktp_view& ktp) {
```
nit: `ktp_view` is cheap to copy
```suggestion
          std::ranges::for_each(ktps, [&](ktp_view ktp) {
```

pgellert commented on src/v/kafka/server/consumer_group_lag_metrics_frontend.h:
```diff
@@ -0,0 +1,63 @@
+    get_partition_offsets(partition_offsets_request req);
+
+private:
+    friend class consumer_group_lag_metrics_service;
```
nit: `consumer_group_lag_metrics_service` is only introduced in a later commit, so this commit wouldn't compile

BenPope replied:
> nit: `consumer_group_lag_metrics_service` is only introduced in a later commit, so this commit wouldn't compile

It does compile, but fair point. I added it to `fwd.h`

pgellert commented on src/v/kafka/server/consumer_group_lag_metrics_service.cc:
```diff
@@ -0,0 +1,32 @@
+  partition_offsets_request req, ::rpc::streaming_context&) {
+    vlog(
+      klog.warn,
+      "consumer_group_lag_metrics_service::partition_offsets: {}",
```
nit: maybe it's worth adding that this is the count of topics, not topic-partitions
```suggestion
      "consumer_group_lag_metrics_service::partition_offsets: {} topics",
```

pgellert commented on src/v/kafka/server/consumer_group_lag_metrics_service.cc:
```diff
@@ -0,0 +1,32 @@
+consumer_group_lag_metrics_service::partition_offsets(
+  partition_offsets_request req, ::rpc::streaming_context&) {
+    vlog(
+      klog.warn,
```
I assume this was `warn` only for debugging:
```suggestion
      klog.debug,
```

pgellert commented on src/v/redpanda/application.cc:
```diff
@@ -3083,6 +3084,11 @@ void application::start_runtime_services(
                   smp_service_groups.datalake_sg(),
                   &_datalake_coordinator_fe));
           }
+          runtime_services.push_back(
+            std::make_unique<kafka::consumer_group_lag_metrics_service>(
+              sched_groups.cluster_sg(),
+              smp_service_groups.cluster_smp_sg(),
+              std::ref(_consumer_group_lag_metrics_frontend)));
```
RPC's are started before the kafka server ([application.cc#L2779-L2808](https://github.com/redpanda-data/redpanda/blob/f7fe9d19d6d5cbbc55966eef75bf10d4873a8c28/src/v/redpanda/application.cc#L2779-L2808)). That seems okay as long as we check in `consumer_group_lag_metrics_service::partition_offsets` if `_frontend.local_is_initialized()`.

pgellert commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -127,29 +134,53 @@ ss::future<> group_manager::start() {
+        };
+
+        _lag_metrics_timer.set_callback([this, set_lag_timer]() {
+            ssx::spawn_with_gate(_gate, [this, set_lag_timer] -> ss::future<> {
```
nit:
```suggestion
            ssx::background = ssx::spawn_with_gate_then(_gate, [this, set_lag_timer] -> ss::future<> {
```

BenPope replied:
Isn't that exactly what `spawn_with_gate` does?

pgellert commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -127,29 +134,53 @@ ss::future<> group_manager::start() {
+        };
+
+        _lag_metrics_timer.set_callback([this, set_lag_timer]() {
+            ssx::spawn_with_gate(_gate, [this, set_lag_timer] -> ss::future<> {
```
If this co-routine lambda safe? Why not use continuation-style in the lambda?

BenPope replied:
> If this co-routine lambda safe? Why not use continuation-style in the lambda?

Changed

pgellert commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1881,4 +1913,59 @@ group_manager::get_group_producers_locally(
+                topic_it != part_offsets.data.end()) {
+                if (auto partition_it = topic_it->second.find(tp.partition);
+                    partition_it != topic_it->second.end()) {
+                    lag part_lag{
+                      static_cast<lag>(partition_it->second)
+                      - offset_cast(group_offset->metadata.offset)};
```
Can this ever become negative? For example if all of the records are deleted from a topic?

BenPope replied:
I've added a `std::max(lag, 0)`

pgellert commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1881,4 +1913,59 @@ group_manager::get_group_producers_locally(
+        co_return;
+    }
+
+    auto holder = _gate.hold();
```
I think we're already holding the gate here from the `spawn_with_gate` that triggered the call to `collect_consumer_lag_metrics`, no?
```suggestion
```

pgellert commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1881,4 +1913,59 @@ group_manager::get_group_producers_locally(
+            request.data[tp.topic].insert(tp.partition);
+            group_offset[tp.topic][tp.partition] = offset_cast(
+              meta->metadata.offset);
+            ;
```
```suggestion
```

pgellert commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1881,4 +1913,59 @@ group_manager::get_group_producers_locally(
+
+    using lag = size_t;
+    partition_offsets_request request;
+    chunked_hash_map<kafka::group_id, partition_offsets_reply::offsets>
+      group_offsets;
```
This `group_offsets` map is just used to check if any groups exist, right? Could it be removed and replaced with a boolean / direct check on `_groups`?

I'm wondering if your original idea here was to use the pre-RPC-call group offsets for the metric instead of the post-RPC-call, but post-RPC-call makes sense to me to keep the lag >0 under normal circumstances.

BenPope replied:
> This `group_offsets` map is just used to check if any groups exist, right? Could it be removed and replaced with a boolean / direct check on `_groups`?

Well, it's needed to gather the set of TPs for committed offsets.

> I'm wondering if your original idea here was to use the pre-RPC-call group offsets for the metric instead of the post-RPC-call, but post-RPC-call makes sense to me to keep the lag >0 under normal circumstances.

That indeed was the plan. Using the offsets from the pre-RPC call ensures the lag isn't negative, no?

pgellert reviewed (commented):
I left some minor comments/nits, but overall looks great so far.

BenPope committed:
kafka/lag_metrics: Add configuration

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
kafka/lag_metrics: Add tests

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/b3fb9ca14429b19eebe60678ada8c6f11cb4b5e6..60ce7c3f22b71a7e0cad0b91bf444f5a1ab7dc02)
* Added `consumer_group_lag_collection_interval`
* Moved types and functionality from `_util.h` to `_frontend.cc`
* Calculate lag from committed offsets taken before the remote call to get partition HWM
* Added tests for metrics
* A few minor review nits

pgellert commented on src/v/config/configuration.h:
```diff
@@ -197,6 +197,7 @@ struct configuration final : public config_store {
     property<bool> disable_public_metrics;
     property<bool> aggregate_metrics;
     property<std::vector<ss::sstring>> enable_consumer_group_metrics;
+    property<std::chrono::seconds> consumer_group_lag_collection_interval;
```
Should we add a `_sec` suffix to the name to clarify the units?

BenPope replied:
> Should we add a `_sec` suffix to the name to clarify the units?

The type doesn't need it, but maybe the name does?

michael-redpanda replied:
yes the name should have `_sec` on it

pgellert replied:
Yeah, I meant the name. I see that we have a bunch of cluster configs with `_ms` and `_sec` in the name, so it might be nice for consistency. And I guess it makes config files a bit easier to read.

BenPope replied:
Done

pgellert commented on tests/rptest/tests/consumer_group_test.py:
```diff
@@ -754,6 +755,203 @@ def get_group_metrics_from_nodes(patterns):
+                logger=self.logger)
+
+        consumers = [create_consumer(i) for i in range(consumer_count)]
+        consumers[0].subscribe(topics)
```
Should this just be `consumers[0]` subscribing the topics or all of them? I guess it works because the test is using `consumer_count = 1`. Would it make sense to simplify the test just to have a single consumer and remove the for loops?

BenPope replied:
> Should this just be `consumers[0]` subscribing the topics or all of them? I guess it works because the test is using `consumer_count = 1`. Would it make sense to simplify the test just to have a single consumer and remove the for loops?

Good spot, I already have changes to use 4 consumers that subscribes them all.

I'd rather keep the complexity to make sure that the internal loops are working.

BenPope replied:
Done

pgellert commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1881,4 +1916,61 @@ group_manager::get_group_producers_locally(
+                }
+                lag part_lag{static_cast<lag>(
+                  std::max(partition_it->second() - offset(), 0L))};
+                lag_metrics.sum += std::max(lag{0}, part_lag);
```
nit: the max against 0 is already done above
```suggestion
                lag part_lag{static_cast<lag>(
                  std::max(partition_it->second() - offset(), 0L))};
                lag_metrics.sum += part_lag;
```

BenPope replied:
Done

pgellert commented on tests/rptest/tests/consumer_group_test.py:
```diff
@@ -754,6 +755,203 @@ def get_group_metrics_from_nodes(patterns):
             c.free()
 
+    @cluster(num_nodes=3)
+    def test_group_lag_metrics(self):
```
🔥

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62770-->
<!--dbd8c02c-3dbf-43b8-81de-181ad4f67cf6-->
<!--title-->
## Retry command for [Build#62770](https://buildkite.com/redpanda/redpanda/builds/62770)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/consumer_group_test.py::ConsumerGroupTest.test_group_recovery
```
<!--/body-->

BenPope commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/60ce7c3f22b71a7e0cad0b91bf444f5a1ab7dc02..a0b0f6eb5e64a0637a46af3789afe9f804a85616)
* Test:
  * Test with 4 consumers
  * Produce a few more messages to ensure every partition is read from
* Minor review nits

michael-redpanda commented on src/v/config/configuration.cc:
```diff
@@ -605,6 +605,13 @@ configuration::configuration()
+      "How oftem to run the collection loop when enable_consumer_group_metrics "
+      "contains consumer_lag",
+      {.needs_restart = needs_restart::no},
+      60s)
```
nit: should this be a tunable?

BenPope replied:
> nit: should this be a tunable?

I made it so.

pgellert reviewed (dismissed):
looks great

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62776-->
<!--83b1a36f-13b0-472a-b670-8a4e814c4655-->
<!--title-->
## Retry command for [Build#62776](https://buildkite.com/redpanda/redpanda/builds/62776)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/consumer_group_test.py::ConsumerGroupTest.test_group_recovery
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62789-->
<!--383b2b64-322c-42bd-864f-2ba7584cb0f9-->
<!--e1e8a986-6635-475b-92b6-7202765eef2a-->
<!--title-->
## Retry command for [Build#62789](https://buildkite.com/redpanda/redpanda/builds/62789)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/consumer_group_test.py::ConsumerGroupTest.test_group_recovery
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--489a7eda-1178-46db-a494-4ef3fa743050-->
<!--TEST RESULTS-->
<!--d69878c2-fd84-4976-a0c7-865b6f894fb3-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62789">build#62789</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.consumer_group_test.ConsumerGroupTest.test_group_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/62789#019571db-ea75-4c9c-aebd-426e86e2a93d | FAIL | 0/20 |
| rptest.tests.consumer_group_test.ConsumerGroupTest.test_group_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/62789#019571df-71ea-4b6a-bc69-50e464ddd5bb | FAIL | 0/20 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62789#019571df-71eb-45d8-af37-aa98361cba09 | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62789#019571df-71ea-4b6a-bc69-50e464ddd5bb | FLAKY | 1/2 |
| rptest.tests.shadow_indexing_compacted_topic_test.ShadowIndexingCompactedTopicTest.test_upload.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62789#019571db-ea74-44e4-874e-8cf0d740dea5 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62841">build#62841</a></summary>


| test_id                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62841#01957fb9-48c4-4735-856b-42c86dfd3ed4 | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_higher_level_migration_api                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62841#01957f9f-835d-4c8d-ba2f-482f91395cad | FLAKY         | 1/2      |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62841#01957fb9-48c5-4565-b879-e814ccd184d6 | FLAKY         | 1/2      |
</details>
<!--/body-->

dotnwat commented on src/v/kafka/server/consumer_group_lag_metrics_rpc_types.h:
```diff
@@ -0,0 +1,51 @@
+
+    using offsets = chunked_hash_map<
+      model::topic,
+      chunked_hash_map<model::partition_id, kafka::offset>>;
```
instead of adding an RPC to go collect exactly one specific integer, should `kafka::offsets` be replaced by a versioned serde type? and/or the RPC itself be generalized a bit to "partition metadata". I'm genuinely not sure, but it seems worth the effort to think about it a little bit so we don't have RPC bloat.

BenPope replied:
Maybe I'm not being very creative, but I can't think of anything other than partition hwm that we would need from other nodes when computing consumer lag.

dotnwat replied:
> partition hwm that we would need from other nodes when computing consumer lag.

the point of serde is that you don't have to be creative today; you can avoid the pain later of evolving a non-serde rpc.

dotnwat commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -127,29 +136,54 @@ ss::future<> group_manager::start() {
+
+        _lag_metrics_timer.set_callback([this, set_lag_timer]() {
+            ssx::spawn_with_gate(_gate, [this, set_lag_timer] {
+                return collect_consumer_lag_metrics().then(
+                  [this, set_lag_timer] { set_lag_timer(this); });
```
it looks like you'd probably want to set `set_lag_timer` even if `collect_consumer_lag_metrics` experiences a problem?

BenPope commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/3822ded578f88400cb7f3541003086bc8ca9c677..d8df16afa274b726e18d93d3d934a311eaca4e03)
* Importing `confluent_kafka.TopicPartition` unintentionally shadowed `kafka.structs.TopicPartition` breaking an unrelated test. Instead, use a dict of `TopicPartition` to offset.

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/d8df16afa274b726e18d93d3d934a311eaca4e03..762ab9f1f7eb6ddf37533dca70b4eca2d2d9d295)
* Always `set_lag_timer` even if `collect_consumer_lag_metrics` experiences a problem.

emaxerrno commented:
@BenPope what's the overhead. so N^2 core count of cluster?

BenPope commented:
> @BenPope what's the overhead. so N^2 core count of cluster?

Every core potentially sends a request to every broker, each reply requires gathering across potentially every core (Cores * Brokers).

I could do a scatter gather for the requesting core so that fewer requests are sent (Brokers^2)
